File: examples/backend-node/src/routes/me.ts
import type { VoidhashNodeClient } from "@voidhash/node";

import type { EntitlementsCache } from "../entitlements-cache";
import { requireDistinctId, sendJson } from "../http";
import type { NoteStore } from "../notes";
import type { RouteHandler } from "../server";
import { findPersonByDistinctId } from "../voidhash";

export type MeRouteOptions = {
  readonly voidhash: VoidhashNodeClient;
  readonly entitlements: EntitlementsCache;
  readonly notes: NoteStore;
};

/**
 * `GET /v1/me?distinctId=…` — who the caller is and what they may do.
 *
 * A distinct id Voidhash has never seen is a free user with no grants, not a
 * 500: the client identifies people lazily, so "unknown" is a normal state.
 */
export const createMeRoute = (options: MeRouteOptions): RouteHandler => {
  const { entitlements, notes, voidhash } = options;

  return async (_request, response, url) => {
    const distinctId = requireDistinctId(url);

    const [person, access] = await Promise.all([
      findPersonByDistinctId(voidhash, distinctId),
      entitlements.resolve(distinctId),
    ]);

    sendJson(response, 200, {
      attributes: {
        notes_created: notes.count(distinctId),
        plan: access.hasPro ? "pro" : "free",
      },
      distinctId,
      entitlementsStale: access.stale,
      grants: access.grants,
      person,
    });
  };
};

File: examples/backend-rust/src/routes/me.rs
//! `GET /v1/me` — who the caller is and what they are entitled to.

use axum::extract::State;
use axum::Json;
use serde::Serialize;

use crate::entitlements::{Freshness, Grant, Person};
use crate::error::ApiError;
use crate::routes::DistinctId;
use crate::state::SharedState;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MeResponse {
    distinct_id: String,
    /// `false` when Voidhash has never seen this distinct id.
    known: bool,
    pro: bool,
    person: Option<Person>,
    grants: Vec<Grant>,
    freshness: Freshness,
}

/// `GET /v1/me?distinctId=…`
///
/// An unknown distinct id is a free user with no grants, not a 404.
pub async fn me(
    State(state): State<SharedState>,
    DistinctId(distinct_id): DistinctId,
) -> Result<Json<MeResponse>, ApiError> {
    let resolved = state.entitlements.resolve(&distinct_id).await?;

    Ok(Json(MeResponse {
        distinct_id,
        known: resolved.entitlements.known,
        pro: resolved.entitlements.pro,
        person: resolved.entitlements.person,
        grants: resolved.entitlements.grants,
        freshness: resolved.freshness,
    }))
}
